From f18d1257ab079d086c6861e1c04dd91f9f69d11f Mon Sep 17 00:00:00 2001
From: Mohsen Dolaty <mohsen.dolaty@amd.com>
Date: Mon, 25 Jul 2022 17:15:32 -0500
Subject: [PATCH 1/1] u-boot: add support to report BMC Boot Cause

Ads support to read syscontrol reset status register to check if BMC
booting after AC power on.
Exports the boolean value in u-boot env var, so
that dimm-info.service can Read DIMM SPD.

Signed-off-by: Mohsen Dolaty <mohsen.dolaty@amd.com>
---
 arch/arm/mach-aspeed/ast2600/scu_info.c |  6 ++++++
 common/main.c                           | 13 +++++++++++++
 2 files changed, 19 insertions(+)

diff --git a/arch/arm/mach-aspeed/ast2600/scu_info.c b/arch/arm/mach-aspeed/ast2600/scu_info.c
index fe26f743c0..838b2b9588 100644
--- a/arch/arm/mach-aspeed/ast2600/scu_info.c
+++ b/arch/arm/mach-aspeed/ast2600/scu_info.c
@@ -12,6 +12,9 @@
 /* SoC mapping Table */
 #define SOC_ID(str, rev) { .name = str, .rev_id = rev, }

+/* Sys Scratch reg SCU1FC */
+#define ASPEED_SYS_SCRATCH_1FC 0x1e6e21fc
+
 struct soc_id {
 	const char *name;
 	u64 rev_id;
@@ -217,6 +220,9 @@ void aspeed_print_sysrst_info(void)
 	u32 rest = readl(ASPEED_SYS_RESET_CTRL);
 	u32 rest3 = readl(ASPEED_SYS_RESET_CTRL3);

+	/* Save ASPEED_SYS_RESET_CTRL value in scratch reg */
+	writel(rest, ASPEED_SYS_SCRATCH_1FC);
+
 	if (rest & SYS_PWR_RESET_FLAG) {
 		printf("RST: Power On \n");
 		writel(rest, ASPEED_SYS_RESET_CTRL);
diff --git a/common/main.c b/common/main.c
index 78dd743096..f3b8ee65d8 100644
--- a/common/main.c
+++ b/common/main.c
@@ -7,6 +7,8 @@
 /* #define	DEBUG	*/

 #include <common.h>
+#include <errno.h>
+#include <asm/io.h>
 #include <autoboot.h>
 #include <cli.h>
 #include <console.h>
@@ -21,6 +23,8 @@
 #define EEPROM_ADDR_LEN		2
 #define BOARD_ID_LEN			1
 #define BOARD_ID_BUFF_LEN		2
+/* Sys Scratch reg that holds sys_rst info */
+#define ASPEED_SYS_SCRATCH_1FC 0x1e6e21fc

 #define ONYX_SLT        61  //0x3D
 #define ONYX_1          64  //0x40
@@ -181,6 +185,7 @@ static void run_preboot_environment_command(void)
 void main_loop(void)
 {
 	const char *s;
+	u32 por_rst = readl(ASPEED_SYS_SCRATCH_1FC);

 	bootstage_mark_name(BOOTSTAGE_ID_MAIN_LOOP, "main_loop");

@@ -194,6 +199,14 @@ void main_loop(void)
 	if (IS_ENABLED(CONFIG_UPDATE_TFTP))
 		update_tftp(0UL, NULL, NULL);

+	/* set reset reason env */
+	printf("Scratch register value: 0x%08x\n", por_rst);
+	if (por_rst & (u32)0x1)
+		env_set("por_rst", "true");
+	else
+		env_set("por_rst", "false");
+	env_save();
+
 	/* Read board id from eerpom and set env */
 	set_board_id();
 	set_hostname();
--
2.25.1
